File: c/dag-1/main.c
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>

#define len 1000

int file_to_buffers(int *left, int *right);
int compare(const void *l, const void *r);

struct Counter {
  int num;
  int count;
};

int main() {
  int left[len] = {0};
  int right[len] = {0};
  if (file_to_buffers(left, right)) {
    return 1;
  }

  qsort(left, len, sizeof(left[0]), compare);
  qsort(right, len, sizeof(right[0]), compare);

  int sum = 0;
  for (int i = 0; i < len; i++) {
    sum += left[i] < right[i] ? right[i] - left[i] : left[i] - right[i];
  }

  printf("The total sum is: %d", sum);

  // Part 1: Done

  struct Counter counter[len] = {0};

  return 0;
}

int file_to_buffers(int *left, int *right) {
  FILE *in = fopen("input.txt", "r");
  if (in == NULL) {
    printf("Couldn't open input.txt");
    return 1;
  }

  int i = 0;
  while (fscanf(in, "%d %d", left + i, right + i) != EOF) {
    i++;
  }

  fclose(in);
  return 0;
}

int compare(const void *l, const void *r) { return (*(int *)l - *(int *)r); }

File: rust/dag-1/src/main.rs
use std::{
    collections::HashMap,
    fs::File,
    io::{BufRead, BufReader},
};

fn main() {
    let file = File::open("input.txt").expect("Couldnt open file");
    let lines = BufReader::new(file).lines().map(|l| l.unwrap());

    let (mut left, mut right): (Vec<i32>, Vec<i32>) = lines
        .map(|line| {
            let mut split = line.split_whitespace();
            (
                split.next().unwrap().parse::<i32>().unwrap(),
                split.next().unwrap().parse::<i32>().unwrap(),
            )
        })
        .unzip();

    left.sort();
    right.sort();

    let sum = left
        .iter()
        .zip(right.iter())
        .fold(0, |acc, (left_num, right_num)| {
            acc + (left_num - right_num).abs()
        });

    // Part 1 Done
    println!("{sum}");

    let mut counter_right = HashMap::new();
    for right in right.iter() {
        *counter_right.entry(right).or_insert(0) += 1;
    }

    let mut score = 0;
    for ref num in left {
        score += num * counter_right.get(num).unwrap_or(&0);
    }

    // Part 2 Done
    println!("{score}");
}
